concrete/src/shortints/types/static.rs

```diff
@@ -278,3 +320,21 @@ impl Default for FheUint4Parameters {
         Self::with_carry_4()
     }
 }
+
+#[cfg(test)]
+mod test {
+    #[cfg(feature = "serde")]
+    use super::*;
+
+    #[cfg(feature = "serde")]
+    #[test]
+    fn can_serialize_deserialize_shortint_id() {
+        let id = ShorIntId::<2>;
+        let mut cursor = std::io::Cursor::new(Vec::<u8>::new());
+        bincode::serialize_into(&mut cursor, &id).unwrap();
+        cursor.set_position(0);
+        let id2: ShorIntId<2> = bincode::deserialize_from(cursor).unwrap();
+
+        assert_eq!(id, id2);
```

Comment:
does this actually serialize something ?

Or said differently the const generic serialization actually stores the 2 in the cursor in the test ?

if we change `ShortIntId<2>` to `ShortIntId<3>` line 336 we will get a serialization error or an assert_eq failure line 338 ?

Reply:
If we change the const number on line 336 the code won't compile because in the assert, we would now be comparing two different types (`ShortIntId<2>`vs `ShortIntId<3>`).
If we remove that assert the code runs fine 

It's the same behaviour for non-const generics unit structs in bincode:

```rust
  #[derive(Serialize)]
  struct UnitStructA;
  #[derive(Deserialize)]
  struct UnitStructB;

  let a = UnitStructA;
  let mut cursor = std::io::Cursor::new(Vec::<u8>::new());
  bincode::serialize_into(&mut cursor, &a).unwrap();
  cursor.set_position(0);
  let b: UnitStructB = bincode::deserialize_from(cursor).unwrap();
```


We could save the const MESSAGE_BITS value as if it where a field during the serialization,
allow a check during deserialization.

```rust

#[cfg(feature = "serde")]
impl<const MESSAGE_BITS: u8> Serialize for ShorIntId<MESSAGE_BITS> {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        let mut state = serializer.serialize_struct("ShorIntId", 0)?;
        state.serialize_field("MESSAGE_BITS", &MESSAGE_BITS)?;
        state.end()
    }
}

#[cfg(feature = "serde")]
impl<'de, const MESSAGE_BITS: u8> Deserialize<'de> for ShorIntId<MESSAGE_BITS> {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        struct Visitor<const MESSAGE_BITS: u8>;

        impl<const MESSAGE_BITS: u8> Visitor<MESSAGE_BITS> {
            const MESSAGE_BITS_FIELD_NAME: &'static str = "MESSAGE_BITS";

            /// Returns a ShorIntId<MESSAGE_BITS> if `parsed_message_bits==MESSAGE_BITS`
            /// Otherwise a deserialization error is returned
            fn try_create_id<E: serde::de::Error>(
                self,
                parsed_message_bits: u8,
            ) -> Result<ShorIntId<MESSAGE_BITS>, E> {
                if MESSAGE_BITS != parsed_message_bits {
                    let displayed = format!("ShorIntId<{}>", parsed_message_bits);
                    let unexpected = serde::de::Unexpected::Other(&displayed);
                    Err(E::invalid_value(unexpected, &self))
                } else {
                    Ok(ShorIntId::<MESSAGE_BITS>)
                }
            }
        }

        impl<'de, const MESSAGE_BITS: u8> serde::de::Visitor<'de> for Visitor<MESSAGE_BITS> {
            type Value = ShorIntId<MESSAGE_BITS>;

            fn expecting(&self, formatter: &mut Formatter) -> std::fmt::Result {
                write!(formatter, "ShorIntId<{}>", MESSAGE_BITS)
            }

            fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
            where
                A: SeqAccess<'de>,
            {
                let message_bits: u8 = seq
                    .next_element()?
                    .ok_or_else(|| serde::de::Error::invalid_length(0, &self))?;
                self.try_create_id(message_bits)
            }

            fn visit_map<A>(self, mut map: A) -> Result<Self::Value, A::Error>
            where
                A: MapAccess<'de>,
            {
                let mut message_bits = None;
                while let Some(key) = map.next_key::<&str>()? {
                    if key == Self::MESSAGE_BITS_FIELD_NAME {
                        if message_bits.is_some() {
                            return Err(serde::de::Error::duplicate_field(
                                Self::MESSAGE_BITS_FIELD_NAME,
                            ));
                        }
                        message_bits = Some(map.next_value()?);
                    }
                }

                let message_bits = message_bits.ok_or_else(|| {
                    serde::de::Error::missing_field(Self::MESSAGE_BITS_FIELD_NAME)
                })?;
                self.try_create_id(message_bits)
            }
        }

        deserializer.deserialize_struct(
            "ShorIntId",
            &[Visitor::<MESSAGE_BITS>::MESSAGE_BITS_FIELD_NAME],
            Visitor::<MESSAGE_BITS>,
        )
    }
}


    #[cfg(feature = "serde")]
    #[test]
    fn can_serialize_deserialize_shortint_id() {
        use serde::de::Error;

        let id = ShorIntId::<2>;
        let mut cursor = std::io::Cursor::new(Vec::<u8>::new());
        bincode::serialize_into(&mut cursor, &id).unwrap();
        cursor.set_position(0);
        let result: Result<ShorIntId<3>, _> = bincode::deserialize_from(&mut cursor);
        let expected_msd = "invalid value: ShorIntId<2>, expected ShorIntId<3>";
        assert!(matches!(result.err().as_deref(), Some(bincode::ErrorKind::Custom(msg)) if msg == expected_msd));


        cursor.set_position(0);
        let result: Result<ShorIntId<2>, _> = bincode::deserialize_from(cursor);
        assert!(matches!(result, Ok(ShorIntId::<2>)));
    }

```

Reply:
Ok, I think we have serialization issues a bit everywhere anyways (sadly) to check that we are deserializing what we expect, I'll unblock the PR if you need it

Reply:
I'd be enclined to stick with the version that does not but the message bits into the data
as it seems what most people do when ser/de struct with const generics (the const values are not serialized)

Reply:
Yes all I'm saying is we have had cases (for core used in the compiler e.g.) where we are missing some checks to verify we are deserializing what we expect, at the moment a lot of our serde/infrastructure seems to be lacking those sanity checks and can serialize/deserialize random stuff